<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


Execute the below code cell to install the required libraries


In [2]:
%pip install pandas
%pip install ipython-sql prettytable 

import prettytable

prettytable.DEFAULT = 'DEFAULT'

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.1.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



  Using cached ipython_sql-0.5.0-py3-none-any.whl.metadata (17 kB)
  Using cached prettytable-3.17.0-py3-none-any.whl.metadata (34 kB)
  Using cached sqlparse-0.5.5-py3-none-any.whl.metadata (4.7 kB)
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl.metadata (755 bytes)
Using cached ipython_sql-0.5.0-py3-none-any.whl (20 kB)
Using cached prettytable-3.17.0-py3-none-any.whl (34 kB)
   ---------------------------------------- 0.0/2.1 MB ? eta -:--:--
   ---------------------------------------- 2.1/2.1 MB 20.0 MB/s eta 0:00:00
Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
Using cached sqlparse-0.5.5-py3-none-any.whl (46 kB)

   ---------------------------------------- 0/6 [ipython-genutils]
   ------ --------------------------------- 1/6 [sqlparse]
   ------ --------------------------------- 1/6 [sqlparse]
   ------ --------------------------------- 1/6 [sqlparse]
   ------------- -------------------------- 2/6 [prettytable]
   -------------------- -----------


[notice] A new release of pip is available: 25.1.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
%pip install --upgrade pip

   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 1.8/1.8 MB 13.9 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 25.1.1
    Uninstalling pip-25.1.1:
      Successfully uninstalled pip-25.1.1
Note: you may need to restart the kernel to use updated packages.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [7]:
import pandas as pd
import sqlite3

conn = sqlite3.connect("FinalDB.db")

Load the SQL magic module


In [8]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [10]:
census_df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01")
schools_df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01")
crime_df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01")

census_df.to_sql("CENSUS_DATA", conn, if_exists="replace", index=False)
schools_df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", conn, if_exists="replace", index=False)
crime_df.to_sql("CHICAGO_CRIME_DATA", conn, if_exists="replace", index=False)

533

Establish a connection between SQL magic module and the database `FinalDB.db`


In [11]:
%%sql sqlite:///FinalDB.db
SELECT name FROM sqlite_master WHERE type='table';

Done.


[('CENSUS_DATA',), ('CHICAGO_PUBLIC_SCHOOLS_DATA',), ('CHICAGO_CRIME_DATA',)]

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [16]:
%sql SELECT * FROM PRAGMA_TABLE_INFO('CHICAGO_CRIME_DATA');

 * sqlite:///FinalDB.db
Done.


[(0, 'ID', 'INTEGER', 0, None, 0),
 (1, 'CASE_NUMBER', 'TEXT', 0, None, 0),
 (2, 'DATE', 'TEXT', 0, None, 0),
 (3, 'BLOCK', 'TEXT', 0, None, 0),
 (4, 'IUCR', 'TEXT', 0, None, 0),
 (5, 'PRIMARY_TYPE', 'TEXT', 0, None, 0),
 (6, 'DESCRIPTION', 'TEXT', 0, None, 0),
 (7, 'LOCATION_DESCRIPTION', 'TEXT', 0, None, 0),
 (8, 'ARREST', 'INTEGER', 0, None, 0),
 (9, 'DOMESTIC', 'INTEGER', 0, None, 0),
 (10, 'BEAT', 'INTEGER', 0, None, 0),
 (11, 'DISTRICT', 'INTEGER', 0, None, 0),
 (12, 'WARD', 'REAL', 0, None, 0),
 (13, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (14, 'FBICODE', 'TEXT', 0, None, 0),
 (15, 'X_COORDINATE', 'REAL', 0, None, 0),
 (16, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (17, 'YEAR', 'INTEGER', 0, None, 0),
 (18, 'LATITUDE', 'REAL', 0, None, 0),
 (19, 'LONGITUDE', 'REAL', 0, None, 0),
 (20, 'LOCATION', 'TEXT', 0, None, 0)]

In [17]:
%%sql
SELECT * FROM CHICAGO_CRIME_DATA LIMIT 5;

 * sqlite:///FinalDB.db
Done.


[(3512276, 'HK587712', '2004-08-28', '047XX S KEDZIE AVE', '890', 'THEFT', 'FROM BUILDING', 'SMALL RETAIL STORE', 0, 0, 911, 9, 14.0, 58.0, '6', 1155838.0, 1873050.0, 2004, 41.8074405, -87.70395585, '(41.8074405, -87.703955849)'),
 (3406613, 'HK456306', '2004-06-26', '009XX N CENTRAL PARK AVE', '820', 'THEFT', '$500 AND UNDER', 'OTHER', 0, 0, 1112, 11, 27.0, 23.0, '6', 1152206.0, 1906127.0, 2004, 41.89827996, -87.71640551, '(41.898279962, -87.716405505)'),
 (8002131, 'HT233595', '2011-04-04', '043XX S WABASH AVE', '820', 'THEFT', '$500 AND UNDER', 'NURSING HOME/RETIREMENT HOME', 0, 0, 221, 2, 3.0, 38.0, '6', 1177436.0, 1876313.0, 2011, 41.81593313, -87.62464213, '(41.815933131, -87.624642127)'),
 (7903289, 'HT133522', '2010-12-30', '083XX S KINGSTON AVE', '840', 'THEFT', 'FINANCIAL ID THEFT: OVER $300', 'RESIDENCE', 0, 0, 423, 4, 7.0, 46.0, '6', 1194622.0, 1850125.0, 2010, 41.74366532, -87.56246276, '(41.743665322, -87.562462756)'),
 (10402076, 'HZ138551', '2016-02-02', '033XX W 66TH ST', '820', 'THEFT', '$500 AND UNDER', 'ALLEY', 0, 0, 831, 8, 15.0, 66.0, '6', 1155240.0, 1860661.0, 2016, 41.7734553, -87.70648047, '(41.773455295, -87.706480471)')]

In [22]:
%sql select count(*) from CHICAGO_CRIME_DATA;

 * sqlite:///FinalDB.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [25]:
%sql select * from census_data limit 5;

 * sqlite:///FinalDB.db
Done.


[(1.0, 'Rogers Park', 7.7, 23.6, 8.7, 18.2, 27.5, 23939, 39.0),
 (2.0, 'West Ridge', 7.8, 17.2, 8.8, 20.8, 38.5, 23040, 46.0),
 (3.0, 'Uptown', 3.8, 24.0, 8.9, 11.8, 22.2, 35787, 20.0),
 (4.0, 'Lincoln Square', 3.4, 10.9, 8.2, 13.4, 25.5, 37524, 17.0),
 (5.0, 'North Center', 0.3, 7.5, 5.2, 4.5, 26.2, 57123, 6.0)]

In [30]:
%%sql
select community_area_number, community_area_name, per_capita_income from census_data
where per_capita_income < 11000
order by per_capita_income;

 * sqlite:///FinalDB.db
Done.


[(54.0, 'Riverdale', 8201),
 (30.0, 'South Lawndale', 10402),
 (37.0, 'Fuller Park', 10432),
 (26.0, 'West Garfield Park', 10934)]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [33]:
%%sql
select case_number, description
from chicago_crime_data
where description like '%minor%';

 * sqlite:///FinalDB.db
Done.


[('HL266884', 'SELL/GIVE/DEL LIQUOR TO MINOR'),
 ('HK238408', 'ILLEGAL CONSUMPTION BY MINOR')]

### Problem 4

##### List all kidnapping crimes involving a child?


In [45]:
%%sql
select distinct(primary_type)
from chicago_crime_data;

 * sqlite:///FinalDB.db
Done.


[('THEFT',),
 ('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('OTHER OFFENSE',),
 ('ASSAULT',),
 ('BURGLARY',),
 ('MOTOR VEHICLE THEFT',),
 ('DECEPTIVE PRACTICE',),
 ('ROBBERY',),
 ('CRIMINAL TRESPASS',),
 ('PROSTITUTION',),
 ('WEAPONS VIOLATION',),
 ('PUBLIC PEACE VIOLATION',),
 ('OFFENSE INVOLVING CHILDREN',),
 ('CRIM SEXUAL ASSAULT',),
 ('SEX OFFENSE',),
 ('INTERFERENCE WITH PUBLIC OFFICER',),
 ('GAMBLING',),
 ('LIQUOR LAW VIOLATION',),
 ('ARSON',),
 ('HOMICIDE',),
 ('KIDNAPPING',),
 ('INTIMIDATION',),
 ('STALKING',),
 ('OBSCENITY',),
 ('CONCEALED CARRY LICENSE VIOLATION',),
 ('PUBLIC INDECENCY',),
 ('NON-CRIMINAL',),
 ('OTHER NARCOTIC VIOLATION',),
 ('HUMAN TRAFFICKING',),
 ('NON - CRIMINAL',),
 ('RITUALISM',),
 ('NON-CRIMINAL (SUBJECT SPECIFIED)',),
 ('DOMESTIC VIOLENCE',)]

In [50]:
%%sql
select case_number, description
from chicago_crime_data
where primary_type = 'KIDNAPPING'
and description like '%CHILD%';

 * sqlite:///FinalDB.db
Done.


[('HN144152', 'CHILD ABDUCTION/STRANGER')]

### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [54]:
%%sql
select distinct(location_description)
from chicago_crime_data;

 * sqlite:///FinalDB.db
Done.


[('SMALL RETAIL STORE',),
 ('OTHER',),
 ('NURSING HOME/RETIREMENT HOME',),
 ('RESIDENCE',),
 ('ALLEY',),
 ('PARKING LOT/GARAGE(NON.RESID.)',),
 ('STREET',),
 ('GROCERY FOOD STORE',),
 ('DEPARTMENT STORE',),
 ('BAR OR TAVERN',),
 ('RESIDENTIAL YARD (FRONT/BACK)',),
 ('RESIDENCE PORCH/HALLWAY',),
 ('CTA PLATFORM',),
 ('VACANT LOT/LAND',),
 ('SIDEWALK',),
 ('ATHLETIC CLUB',),
 ('AIRPORT/AIRCRAFT',),
 ('APARTMENT',),
 ('MOVIE HOUSE/THEATER',),
 ('VEHICLE NON-COMMERCIAL',),
 ('RESTAURANT',),
 ('RESIDENCE-GARAGE',),
 ('COMMERCIAL / BUSINESS OFFICE',),
 ('TAVERN/LIQUOR STORE',),
 ('CONSTRUCTION SITE',),
 ('GAS STATION',),
 ('DRUG STORE',),
 ('CTA BUS',),
 ('SCHOOL, PUBLIC, GROUNDS',),
 ('SCHOOL, PUBLIC, BUILDING',),
 ('CHURCH/SYNAGOGUE/PLACE OF WORSHIP',),
 ('OTHER COMMERCIAL TRANSPORTATION',),
 ('CHA APARTMENT',),
 ('WAREHOUSE',),
 ('POLICE FACILITY/VEH PARKING LOT',),
 ('HOSPITAL BUILDING/GROUNDS',),
 ('FACTORY/MANUFACTURING BUILDING',),
 ('CLEANING STORE',),
 ('CTA TRAIN',),
 ('BANK',),
 ('CHA PARKING LOT/GROUNDS',),
 ('LIBRARY',),
 ('CHA HALLWAY/STAIRWELL/ELEVATOR',),
 ('COLLEGE/UNIVERSITY GROUNDS',),
 ('SCHOOL, PRIVATE, BUILDING',),
 ('CONVENIENCE STORE',),
 ('HOUSE',),
 ('GOVERNMENT BUILDING/PROPERTY',)]

In [97]:
%%sql
select distinct(primary_type)
from chicago_crime_data
where location_description like '%SCHOOL%';

 * sqlite:///FinalDB.db
Done.


[('BATTERY',),
 ('CRIMINAL DAMAGE',),
 ('NARCOTICS',),
 ('ASSAULT',),
 ('CRIMINAL TRESPASS',),
 ('PUBLIC PEACE VIOLATION',)]

### Problem 6

##### List the type of schools along with the average safety score for each type.


In [58]:
%%sql
pragma table_info('CHICAGO_PUBLIC_SCHOOLS_DATA');

 * sqlite:///FinalDB.db
Done.


[(0, 'School_ID', 'INTEGER', 0, None, 0),
 (1, 'NAME_OF_SCHOOL', 'TEXT', 0, None, 0),
 (2, 'Elementary, Middle, or High School', 'TEXT', 0, None, 0),
 (3, 'Street_Address', 'TEXT', 0, None, 0),
 (4, 'City', 'TEXT', 0, None, 0),
 (5, 'State', 'TEXT', 0, None, 0),
 (6, 'ZIP_Code', 'INTEGER', 0, None, 0),
 (7, 'Phone_Number', 'TEXT', 0, None, 0),
 (8, 'Link', 'TEXT', 0, None, 0),
 (9, 'Network_Manager', 'TEXT', 0, None, 0),
 (10, 'Collaborative_Name', 'TEXT', 0, None, 0),
 (11, 'Adequate_Yearly_Progress_Made_', 'TEXT', 0, None, 0),
 (12, 'Track_Schedule', 'TEXT', 0, None, 0),
 (13, 'CPS_Performance_Policy_Status', 'TEXT', 0, None, 0),
 (14, 'CPS_Performance_Policy_Level', 'TEXT', 0, None, 0),
 (15, 'HEALTHY_SCHOOL_CERTIFIED', 'TEXT', 0, None, 0),
 (16, 'Safety_Icon', 'TEXT', 0, None, 0),
 (17, 'SAFETY_SCORE', 'REAL', 0, None, 0),
 (18, 'Family_Involvement_Icon', 'TEXT', 0, None, 0),
 (19, 'Family_Involvement_Score', 'TEXT', 0, None, 0),
 (20, 'Environment_Icon', 'TEXT', 0, None, 0),
 (21, 'Environment_Score', 'REAL', 0, None, 0),
 (22, 'Instruction_Icon', 'TEXT', 0, None, 0),
 (23, 'Instruction_Score', 'REAL', 0, None, 0),
 (24, 'Leaders_Icon', 'TEXT', 0, None, 0),
 (25, 'Leaders_Score', 'TEXT', 0, None, 0),
 (26, 'Teachers_Icon', 'TEXT', 0, None, 0),
 (27, 'Teachers_Score', 'TEXT', 0, None, 0),
 (28, 'Parent_Engagement_Icon', 'TEXT', 0, None, 0),
 (29, 'Parent_Engagement_Score', 'TEXT', 0, None, 0),
 (30, 'Parent_Environment_Icon', 'TEXT', 0, None, 0),
 (31, 'Parent_Environment_Score', 'TEXT', 0, None, 0),
 (32, 'AVERAGE_STUDENT_ATTENDANCE', 'TEXT', 0, None, 0),
 (33, 'Rate_of_Misconducts__per_100_students_', 'REAL', 0, None, 0),
 (34, 'Average_Teacher_Attendance', 'TEXT', 0, None, 0),
 (35, 'Individualized_Education_Program_Compliance_Rate', 'TEXT', 0, None, 0),
 (36, 'Pk_2_Literacy__', 'TEXT', 0, None, 0),
 (37, 'Pk_2_Math__', 'TEXT', 0, None, 0),
 (38, 'Gr3_5_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (39, 'Gr3_5_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (40, 'Gr3_5_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (41, 'Gr3_5_Keep_Pace_Math__', 'TEXT', 0, None, 0),
 (42, 'Gr6_8_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (43, 'Gr6_8_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (44, 'Gr6_8_Keep_Pace_Math_', 'TEXT', 0, None, 0),
 (45, 'Gr6_8_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (46, 'Gr_8_Explore_Math__', 'TEXT', 0, None, 0),
 (47, 'Gr_8_Explore_Read__', 'TEXT', 0, None, 0),
 (48, 'ISAT_Exceeding_Math__', 'REAL', 0, None, 0),
 (49, 'ISAT_Exceeding_Reading__', 'REAL', 0, None, 0),
 (50, 'ISAT_Value_Add_Math', 'REAL', 0, None, 0),
 (51, 'ISAT_Value_Add_Read', 'REAL', 0, None, 0),
 (52, 'ISAT_Value_Add_Color_Math', 'TEXT', 0, None, 0),
 (53, 'ISAT_Value_Add_Color_Read', 'TEXT', 0, None, 0),
 (54, 'Students_Taking__Algebra__', 'TEXT', 0, None, 0),
 (55, 'Students_Passing__Algebra__', 'TEXT', 0, None, 0),
 (56, '9th Grade EXPLORE (2009)', 'TEXT', 0, None, 0),
 (57, '9th Grade EXPLORE (2010)', 'TEXT', 0, None, 0),
 (58, '10th Grade PLAN (2009)', 'TEXT', 0, None, 0),
 (59, '10th Grade PLAN (2010)', 'TEXT', 0, None, 0),
 (60, 'Net_Change_EXPLORE_and_PLAN', 'TEXT', 0, None, 0),
 (61, '11th Grade Average ACT (2011)', 'TEXT', 0, None, 0),
 (62, 'Net_Change_PLAN_and_ACT', 'TEXT', 0, None, 0),
 (63, 'College_Eligibility__', 'TEXT', 0, None, 0),
 (64, 'Graduation_Rate__', 'TEXT', 0, None, 0),
 (65, 'College_Enrollment_Rate__', 'TEXT', 0, None, 0),
 (66, 'COLLEGE_ENROLLMENT', 'INTEGER', 0, None, 0),
 (67, 'General_Services_Route', 'INTEGER', 0, None, 0),
 (68, 'Freshman_on_Track_Rate__', 'TEXT', 0, None, 0),
 (69, 'X_COORDINATE', 'REAL', 0, None, 0),
 (70, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (71, 'Latitude', 'REAL', 0, None, 0),
 (72, 'Longitude', 'REAL', 0, None, 0),
 (73, 'COMMUNITY_AREA_NUMBER', 'INTEGER', 0, None, 0),
 (74, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (75, 'Ward', 'INTEGER', 0, None, 0),
 (76, 'Police_District', 'INTEGER', 0, None, 0),
 (77, 'Location', 'TEXT', 0, None, 0)]

In [65]:
%%sql
select "Elementary, Middle, or High School" as School_Type,
AVG(SAFETY_SCORE) as avg_safety_score
from CHICAGO_PUBLIC_SCHOOLS_DATA
group by "Elementary, Middle, or High School";

 * sqlite:///FinalDB.db
Done.


[('ES', 49.52038369304557), ('HS', 49.62352941176471), ('MS', 48.0)]

!!! resulted in only 1 safety rating rather than 3-> because accidentally used single quotations which groups it by a string literal, not the column. everything fell under one group rather than the column !!!

'text' → literal value
"text" → column name (in SQLite)

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [67]:
%%sql
pragma table_info('census_data');

 * sqlite:///FinalDB.db
Done.


[(0, 'COMMUNITY_AREA_NUMBER', 'REAL', 0, None, 0),
 (1, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (2, 'PERCENT_OF_HOUSING_CROWDED', 'REAL', 0, None, 0),
 (3, 'PERCENT_HOUSEHOLDS_BELOW_POVERTY', 'REAL', 0, None, 0),
 (4, 'PERCENT_AGED_16__UNEMPLOYED', 'REAL', 0, None, 0),
 (5, 'PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA', 'REAL', 0, None, 0),
 (6, 'PERCENT_AGED_UNDER_18_OR_OVER_64', 'REAL', 0, None, 0),
 (7, 'PER_CAPITA_INCOME', 'INTEGER', 0, None, 0),
 (8, 'HARDSHIP_INDEX', 'REAL', 0, None, 0)]

In [72]:
%%sql
select community_area_name, percent_households_below_poverty from census_data
order by percent_households_below_poverty desc
limit 5;

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 56.5),
 ('Fuller Park', 51.2),
 ('Englewood', 46.6),
 ('North Lawndale', 43.1),
 ('East Garfield Park', 42.4)]

### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [77]:
%%sql
select community_area_number
from chicago_crime_data
group by community_area_number
order by count(*) desc
limit 1;

 * sqlite:///FinalDB.db
Done.


[(25.0,)]


Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [100]:
%%sql
select community_area_name, hardship_index
from census_data
order by hardship_index desc
limit 1;

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 98.0)]

In [106]:
%%sql
select community_area_name, hardship_index
from census_data
where hardship_index in
(select max(hardship_index) from census_data);

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 98.0)]

Use:

= when comparing to one value

IN when comparing to multiple possible values

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [84]:
%sql select * from chicago_crime_data limit 2;

 * sqlite:///FinalDB.db
Done.


[(3512276, 'HK587712', '2004-08-28', '047XX S KEDZIE AVE', '890', 'THEFT', 'FROM BUILDING', 'SMALL RETAIL STORE', 0, 0, 911, 9, 14.0, 58.0, '6', 1155838.0, 1873050.0, 2004, 41.8074405, -87.70395585, '(41.8074405, -87.703955849)'),
 (3406613, 'HK456306', '2004-06-26', '009XX N CENTRAL PARK AVE', '820', 'THEFT', '$500 AND UNDER', 'OTHER', 0, 0, 1112, 11, 27.0, 23.0, '6', 1152206.0, 1906127.0, 2004, 41.89827996, -87.71640551, '(41.898279962, -87.716405505)')]

In [89]:
%%sql
select community_area_name, count(*) as crime_count
from census_data, chicago_crime_data
where census_data.community_area_number = chicago_crime_data.community_area_number
group by community_area_name
order by crime_count desc
limit 1;

 * sqlite:///FinalDB.db
Done.


[('Austin', 43)]

## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
